plugins/modules/git_config.py

```diff
@@ -197,94 +217,118 @@ def main():
 
-    (rc, out, err) = module.run_command(args, cwd=dir, expand_user_and_vars=False)
     if params['list_all'] and scope and rc == 128 and 'unable to read config file' in err:
         # This just means nothing has been set at the given scope
         module.exit_json(changed=False, msg='', config_values={})
     elif rc >= 2:
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
-        module.fail_json(rc=rc, msg=err, cmd=' '.join(args))
+        module.fail_json(rc=rc, msg=err, cmd=' '.join(list_args))
+
+    old_values = out.rstrip().splitlines()
 
     if params['list_all']:
-        values = out.rstrip().splitlines()
         config_values = {}
-        for value in values:
+        for value in old_values:
             k, v = value.split('=', 1)
             config_values[k] = v
         module.exit_json(changed=False, msg='', config_values=config_values)
     elif not new_value and not unset:
-        module.exit_json(changed=False, msg='', config_value=out.rstrip())
+        module.exit_json(changed=False, msg='', config_value=old_values[0] if old_values else '')
     elif unset and not out:
         module.exit_json(changed=False, msg='no setting to unset')
+    elif new_value in old_values and (len(old_values) == 1 or add_mode == "add"):
+        module.exit_json(changed=False, msg="")
+
+    # Until this point, the git config was just read and in case no change is needed, the module has already exited.
+
+    set_args = list(base_args)
+    if unset:
+        set_args.append("--unset-all")
+        set_args.append(name)
     else:
-        old_value = out.rstrip()
-        if old_value == new_value:
-            module.exit_json(changed=False, msg="")
+        set_args.append("--" + add_mode)
+        set_args.append(name)
+        set_args.append(new_value)
 
     if not module.check_mode:
-        if unset:
-            args.insert(len(args) - 1, "--" + unset)
-            cmd = args
-        else:
-            cmd = args + [new_value]
-        (rc, out, err) = module.run_command(cmd, cwd=dir, ignore_invalid_cwd=False, expand_user_and_vars=False)
+        (rc, out, err) = module.run_command(set_args, cwd=cwd, ignore_invalid_cwd=False, expand_user_and_vars=False)
         if err:
-            module.fail_json(rc=rc, msg=err, cmd=cmd)
+            module.fail_json(rc=rc, msg=err, cmd=set_args)
+
+    if unset:
+        after_values = []
+    elif add_mode == "add":
+        after_values = old_values + [new_value]
+    else:
+        after_values = [new_value]
 
     module.exit_json(
         msg='setting changed',
         diff=dict(
-            before_header=' '.join(args),
-            before=old_value + "\n",
-            after_header=' '.join(args),
-            after=(new_value or '') + "\n"
+            before_header=' '.join(set_args),
+            before=build_diff_value(old_values),
+            after_header=' '.join(set_args),
+            after=build_diff_value(after_values),
```

Comment:
This is still a breaking change in the diff. Should this also be reset to keep compatibility?

Reply:
I think this is OK, as long as it is documented in the changelog.

Reply:
I added another changelog fragement which explicitly mentions that the diff is changed.

Reply:
(btw: Should I resolve discussions or do you want to do that when you are happy with the changes)

Reply:
I personally think it's better for the reviewer to resolve them. Unfortunately GitHub only allows the PR creater to resolve and folks with sufficient permissions to resolve, so if a "regular" module maintainer reviews something, they cannot resolve. Fortunately, I do have enough permissions, so I can do it ;)